### PR TITLE
Converted txt guild experience database to yaml

### PR DESCRIPTION
**Pull request by Atemo**
* **Addressed Issue(s)**: -

* **Server Mode**: both

* **Description of Pull Request**: 

This PR is a suggestion of format to convert exp_guild txt to yaml.

### CODE REVIEW

File: src/char/int_guild.cpp [hunk 1]
```diff
@@ -757,23 +741,62 @@ int inter_guild_CharOffline(uint32 char_id, int guild_id)
...
+
+	if (!this->asUInt16(node, "Level", level))
+		return 0;
+
+	if (level == 0) {
+		this->invalidWarning(node, "The guild level minimum is 1.\n");
+		return 0;
+	}
+	if (level > MAX_GUILDLEVEL)
+		return 0;
```
Comment (aleos89) [inline src/char/int_guild.cpp:763]:
```suggestion
	if (level > MAX_GUILDLEVEL) {
		this->invalidWarning(node["Level"], "Guild level %d exceeds maximum level %d, skipping.\n", level, MAX_GUILDLEVEL);
		return 0;
	}
```

File: src/char/int_guild.cpp [hunk 2]
```diff
@@ -757,23 +741,62 @@ int inter_guild_CharOffline(uint32 char_id, int guild_id)
...
+	}
+
+	uint16 level;
+
+	if (!this->asUInt16(node, "Level", level))
+		return 0;
+
+	if (level == 0) {
+		this->invalidWarning(node, "The guild level minimum is 1.\n");
```
Comment (aleos89) [inline src/char/int_guild.cpp:759]:
```suggestion
		this->invalidWarning(node, "The minimum guild level is 1.\n");
```

File: db/exp_guild.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+###########################################################################
+# Guild Experience Database
+###########################################################################
+#
+# Settings
```
Comment (aleos89) [inline db/exp_guild.yml:22]:
```suggestion
# Guild Experience Settings
```
Just to keep in line with other databases.

File: db/import-tmpl/exp_guild.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+###########################################################################
+# Guild Experience Database
+###########################################################################
+#
+# Settings
```
Comment (aleos89) [inline db/import-tmpl/exp_guild.yml:22]:
```suggestion
# Guild Experience Settings
```
Just to keep in line with other databases.

File: db/pre-re/exp_guild.yml [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+###########################################################################
+# Guild Experience Database
+###########################################################################
+#
+# Settings
```
Comment (aleos89) [inline db/pre-re/exp_guild.yml:22]:
```suggestion
# Guild Experience Settings
```
Just to keep in line with other databases.

File: db/re/exp_guild.yml [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+###########################################################################
+# Guild Experience Database
+###########################################################################
+#
+# Settings
```
Comment (aleos89) [inline db/re/exp_guild.yml:22]:
```suggestion
# Guild Experience Settings
```
Just to keep in line with other databases.

File: doc/yaml/db/exp_guild.yml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+###########################################################################
+# Guild Experience Database
+###########################################################################
+#
+# Settings
```
Comment (aleos89) [inline doc/yaml/db/exp_guild.yml:5]:
```suggestion
# Guild Experience Settings
```
Just to keep in line with other databases.

File: src/char/int_guild.cpp [hunk 8]
```diff
@@ -859,14 +880,14 @@ int guild_calcinfo(struct guild *g)
 
 	if(g->guild_lv<=0)
 		g->guild_lv = 1;
-	nextexp = guild_nextexp(g->guild_lv);
+	nextexp = guild_exp_db.get_nextexp(g->guild_lv);
 
 	// Consume guild exp and increase guild level
 	while(g->exp >= nextexp && nextexp > 0){	//fixed guild exp overflow [Kevin]
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:886]:
```suggestion
	while(g->exp >= g->next_exp && g->next_exp > 0){
```

File: src/char/int_guild.cpp [hunk 9]
```diff
@@ -859,14 +880,14 @@ int guild_calcinfo(struct guild *g)
...
 		g->exp-=nextexp;
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:887]:
```suggestion
		g->exp-=g->next_exp;
```

File: src/char/int_guild.cpp [hunk 10]
```diff
@@ -859,14 +880,14 @@ int guild_calcinfo(struct guild *g)
...
 		g->guild_lv++;
 		g->skill_point++;
-		nextexp = guild_nextexp(g->guild_lv);
+		nextexp = guild_exp_db.get_nextexp(g->guild_lv);
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:890]:
```suggestion
		g->next_exp = guild_exp_db.get_nextexp(g->guild_lv);
```

File: src/char/int_guild.cpp [hunk 11]
```diff
@@ -859,14 +880,14 @@ int guild_calcinfo(struct guild *g)
...
 	}
 
 	// Save next exp step
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:893]: Cant make a suggestion here, but these two lines could be removed with the changes above.

File: src/char/int_guild.cpp [hunk 12]
```diff
@@ -757,23 +741,64 @@ int inter_guild_CharOffline(uint32 char_id, int guild_id)
...
 	castle_db = idb_alloc(DB_OPT_RELEASE_DATA);
 
-	//Read exp file
-	for(i = 0; i<ARRAYLENGTH(filename); i++){
-		sv_readdb(db_path, filename[i], ',', 1, 1, MAX_GUILDLEVEL, exp_guild_parse_row, i > 0);
-	}
-
+	guild_exp_db.load();
+	if (guild_exp_db.size() < (MAX_GUILDLEVEL-1))
+		ShowWarning("Not enough entries for exp_guild, found %" PRIuPTR ", need at least %d.\n", guild_exp_db.size(), (MAX_GUILDLEVEL-1));	// 49 entries for pre-renewal, 50 for renewal
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:798]: Should be moved to loadingFinished.
Also check should be done by checking each level just to be save. :-)

File: src/char/int_guild.cpp [hunk 13]
```diff
@@ -757,23 +741,64 @@ int inter_guild_CharOffline(uint32 char_id, int guild_id)
...
+	uint16 level;
+
+	if (!this->asUInt16(node, "Level", level))
+		return 0;
+
+	if (level == 0) {
+		this->invalidWarning(node, "The minimum guild level is 1.\n");
+		return 0;
+	}
+	if (level > MAX_GUILDLEVEL) {
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:762]:
```suggestion
	if (level >= MAX_GUILDLEVEL) {
```

File: src/tool/csv2yaml.cpp [hunk 14]
```diff
@@ -3451,3 +3463,20 @@ static bool mob_parse_row_chatdb(char* fields[], int columns, int current) {
...
+static bool exp_guild_parse_row(char* split[], int column, int current) {
+	t_exp exp = strtoull(split[0], nullptr, 10);
+
+	if (exp > MAX_GUILD_EXP) {
+		ShowError("exp_guild: Invalid exp %" PRIu64 " at line %d, exceeds max of %" PRIu64 "\n", exp, current, MAX_GUILD_EXP);
+		return false;
+	}
+
+	body << YAML::BeginMap;
+	body << YAML::Key << "Level" << YAML::Value << (current+1);
```
Comment (Lemongrass3110) [inline src/tool/csv2yaml.cpp:3477]: Isnt "current" the line number here?
Or does it ignore the empty lines / comments here?

File: src/char/int_guild.hpp [hunk 15]
```diff
@@ -25,8 +27,24 @@ enum e_guild_action : uint32 {
 struct guild;
 struct guild_castle;
 
+struct s_guild_exp_db {
+	uint16 level;
+	uint64 exp;
```
Comment (Lemongrass3110) [inline src/char/int_guild.hpp:32]:
```suggestion
	t_exp exp;
```

File: src/char/int_guild.cpp (hunk 12 above)
Comment (Atemo) [reply src/char/int_guild.cpp:798]: There are 49 entries in exp_guild.txt in pre-renewal folder (== guild max level is 50), 50 for renewal  (== guild max level is 51 regardless of MAX_GUILDLEVEL) but the value of MAX_GUILDLEVEL is the same for both
```
#define MAX_GUILDLEVEL 50 ///Max Guild level
```
I'm crazy or there is something wrong on the txt version.
Do you know the max guild level on KRO ?

File: src/tool/csv2yaml.cpp (hunk 14 above)
Comment (Atemo) [reply src/tool/csv2yaml.cpp:3477]: It seems 'current' ignores the empty lines and comments

File: src/map/map-server.vcxproj [hunk 16]
```diff
@@ -307,6 +307,8 @@
     <Copy SourceFiles="$(SolutionDir)db\import-tmpl\elemental_skill_db.txt" DestinationFolder="$(SolutionDir)db\import\" ContinueOnError="true" Condition="!Exists('$(SolutionDir)db\import\elemental_skill_db.txt')" />
     <Copy SourceFiles="$(SolutionDir)db\import-tmpl\exp_guild.txt" DestinationFolder="$(SolutionDir)db\import\" ContinueOnError="true" Condition="!Exists('$(SolutionDir)db\import\exp_guild.txt')" />
     <Copy SourceFiles="$(SolutionDir)db\import-tmpl\exp_homun.yml" DestinationFolder="$(SolutionDir)db\import\" ContinueOnError="true" Condition="!Exists('$(SolutionDir)db\import\exp_homun.yml')" />
+    <Copy SourceFiles="$(SolutionDir)db\import-tmpl\exp_guild.yml" DestinationFolder="$(SolutionDir)db\import\" ContinueOnError="true" Condition="!Exists('$(SolutionDir)db\import\exp_guild.yml')" />
+    <Copy SourceFiles="$(SolutionDir)db\import-tmpl\exp_homun.txt" DestinationFolder="$(SolutionDir)db\import\" ContinueOnError="true" Condition="!Exists('$(SolutionDir)db\import\exp_homun.txt')" />
```
Comment (Lemongrass3110) [inline src/map/map-server.vcxproj:311]:
```suggestion
```

File: src/char/int_guild.cpp [hunk 17]
```diff
@@ -757,23 +741,68 @@ int inter_guild_CharOffline(uint32 char_id, int guild_id)
...
+		this->invalidWarning(node["Level"], "Guild level %d exceeds maximum level %d, skipping.\n", level, MAX_GUILDLEVEL);
+		return 0;
+	}
+
+	uint64 exp;
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:767]:
```suggestion
	t_exp exp;
```

File: src/char/int_guild.cpp [hunk 18]
```diff
@@ -757,23 +741,68 @@ int inter_guild_CharOffline(uint32 char_id, int guild_id)
...
+
+	if (!this->asUInt64(node, "Exp", exp))
+		return 0;
+
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:771]: Do we need to check for MAX_GUILD_EXP here?

**Review by Lemongrass3110 — APPROVED** [review]

File: src/char/int_guild.cpp (hunk 12 above)
Comment (Lemongrass3110) [reply src/char/int_guild.cpp]: No idea about pre-renewal anymore, but for renewal this is 50 as far as I know.

File: src/char/int_guild.cpp (hunk 13 above)
Comment (Atemo) [reply src/char/int_guild.cpp:762]: I believe you were right, I commit that

**Review by aleos89 — APPROVED** [review]

File: src/char/int_guild.cpp [hunk 19]
```diff
@@ -757,23 +741,73 @@ int inter_guild_CharOffline(uint32 char_id, int guild_id)
...
+	return 1;
+}
+
+GuildExpDatabase guild_exp_db;
+
+void GuildExpDatabase::loadingFinished() {
+	for (uint16 level = 1; level < MAX_GUILDLEVEL; level++) {
+		if (this->get_nextexp(level) == 0)
+			ShowError("Missing experience for level %d in exp_guild.yml\n", level);
```
Comment (Lemongrass3110) [inline src/char/int_guild.cpp:798]:
```suggestion
			ShowError("Missing experience for guild level %d.\n", level);
```

**Review by Lemongrass3110 — APPROVED** [review]

**Pull request merged**